framework/include/base/Assembly.h

```diff
@@ -672,6 +672,43 @@ class Assembly
 
   /// Temporary work vector to keep from reallocating it
   std::vector<dof_id_type> _temp_dof_indices;
+
+public:
+  /**
+   * We can't add NodalBC values to the Jacobian (or preconditioning)
+   * matrix at the time they are computed -- we instead need to
+   * overwrite an entire row of values with the Jacobian associated to
+   * the NodalBC, which may be coupled to other variables, *after*
+   * matrix assembly is finished.  Hence, the Assembly object
+   * maintains a reference to this data which NodalBCs can write to,
+   * and which can be used later to properly set NodalBC jacobians.
+   * We use numeric_index_type for the index arrays (rather than
+   * dof_id_type) since that is what the SparseMatrix interface uses,
+   * but at the time of this writing, those two types are equivalent.
+   */
+  class CachedNodalBCData
+  {
+  public:
+    // Clears the cached data arrays
+    void clear()
+    {
+      cached_nodal_bc_vals.clear();
```

Comment:
You may want to explicitly remember the `capacity()` of these vectors and call `reserve()` on them after clearing them.

`clear()` is not guaranteed to keep the capacity in tact... and you're using `push_back()` to fill these things in - so it may be a good idea to make sure that `push_back()` is only slow the first time these BCs are computed...